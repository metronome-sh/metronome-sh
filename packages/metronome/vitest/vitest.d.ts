import type { Assertion, AsymmetricMatchersContaining } from "vitest";

interface CustomMatchers<R = unknown> {
  toHaveBeenEventuallyCalled(timeout?: number): Promise<R>;
  toHaveBeenEventuallyCalledWith(...expectedArgs: any[]): Promise<R>;
}

declare module "vitest" {
  interface Assertion<T = any> extends CustomMatchers<T> {}
  interface AsymmetricMatchersContaining extends CustomMatchers {}
}
