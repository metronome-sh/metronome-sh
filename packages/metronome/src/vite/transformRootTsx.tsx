import MagicString from "magic-string";
import { parse } from "@typescript-eslint/typescript-estree";
import { walk } from "./walk";

export function transformRootTsx({ code, id }: { code: string; id: string }) {
  const ast = parse(code, {
    jsx: true,
    useJSXTextNode: true,
    loc: true,
    range: true,
  });

  const magicString = new MagicString(code, { filename: id });

  magicString.prepend('import { withMetronome } from "metronome-sh/react";\n');

  walk(ast, (node) => {
    if (node.type === "ExportDefaultDeclaration" && node.declaration) {
      const [start, end] = node.declaration.range;
      magicString.overwrite(start, end, `withMetronome(${code.substring(start, end)});`);
    }
  });

  const rootTsx = magicString.toString();

  return {
    code: rootTsx,
    map: new MagicString(rootTsx).generateMap({
      includeContent: true,
      source: id,
      file: id,
    }),
  };
}
