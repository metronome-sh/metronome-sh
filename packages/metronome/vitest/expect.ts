import { expect, vi } from "vitest";

expect.extend({
  async toHaveBeenEventuallyCalled(received) {
    const { isNot } = this;
    const timeout = 2000;
    let pass = false;
    try {
      await vi.waitUntil(() => received.mock.calls.length, { timeout });
      pass = true;
    } catch (error) {
      pass = false;
    }

    return {
      pass,
      message: () => `expected mock function ${isNot ? "not " : ""}to have been called`,
    };
  },

  async toHaveBeenEventuallyCalledWith(received, ...expectedArgs) {
    const timeout = 2000;
    const { equals } = this;
    await vi.waitUntil(
      () => {
        return (
          received.mock.calls.length &&
          received.mock.calls.some((args) => equals(args, expectedArgs))
        );
      },
      { timeout }
    );

    return {
      pass: true,
      message: () => `Expected function to have been called with ${expectedArgs} but it was not.`,
      actual: received,
      expected: expectedArgs,
    };
  },
});
