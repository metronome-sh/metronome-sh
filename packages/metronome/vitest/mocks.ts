import { Mock, vi } from "vitest";

export const onMockRequest = vi.fn() as Mock;
