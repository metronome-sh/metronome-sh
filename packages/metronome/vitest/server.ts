import { setupServer } from "msw/node";
import { http, HttpResponse } from "msw";
import { onMockRequest } from "./mocks";

export const handlers = [
  http.post("https://metrics.metronome.sh/v4/process", () => {
    return new HttpResponse(null, { status: 200 });
  }),
];

export const server = setupServer(...handlers);

server.events.on("request:end", async ({ request }) => {
  if (!request.body) return onMockRequest(request.body);

  const reader = request.body.getReader();
  const decoder = new TextDecoder("utf-8");
  let body = "";
  let done, value;

  while (!done) {
    ({ done, value } = await reader.read());
    if (!done) {
      body += decoder.decode(value, { stream: true });
    }
  }

  body += decoder.decode();

  // Uncomment to debug
  // console.log("request:end \n", body);

  try {
    const json = JSON.parse(body);
    onMockRequest(json);
  } catch (error) {
    onMockRequest(body);
  }
});
