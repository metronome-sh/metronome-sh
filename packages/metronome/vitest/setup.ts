import "./expect";
import { server } from "./server";
import { beforeAll, afterAll, afterEach, expect, vi, beforeEach } from "vitest";
import { onMockRequest } from "./mocks";

beforeAll(() => {
  server.listen({ onUnhandledRequest: "error" });
});
beforeEach(() => {
  vi.useFakeTimers();
  vi.setSystemTime(new Date(1988, 4, 13));

  const hrtimeSpy = vi.spyOn(process.hrtime, "bigint");
  hrtimeSpy.mockImplementation(() => 0n);
  return () => {
    hrtimeSpy.mockRestore();
  };
});
afterAll(() => server.close());
afterEach(() => {
  server.resetHandlers();
  onMockRequest.mockClear();
});
