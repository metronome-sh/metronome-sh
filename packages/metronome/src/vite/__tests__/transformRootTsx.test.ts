import { describe, expect, it } from "vitest";
import { transformRootTsx } from "../transformRootTsx";
import fs from "fs";
import path from "path";
describe("transformRootTsx", () => {
  it.each([
    "__templates__/simple-root.txt",
    "__templates__/default-template-root.txt",
    "__templates__/complex-root.txt",
  ])("should transform %s", (filename) => {
    const content = fs.readFileSync(path.resolve(__dirname, filename), "utf-8");

    const { code, map } = transformRootTsx({
      code: content,
      id: filename,
    });

    expect(code).toMatchSnapshot();
    expect(map).toMatchSnapshot();
  });
});
