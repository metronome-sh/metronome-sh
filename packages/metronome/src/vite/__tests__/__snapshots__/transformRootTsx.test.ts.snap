// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transformRootTsx > should transform __templates__/complex-root.txt 1`] = `
"import { withMetronome } from "metronome-sh/react";
import type { LoaderFunctionArgs } from "@remix-run/cloudflare";
import { json } from "@remix-run/cloudflare";
import {
  Links,
  Meta,
  Outlet,
  Scripts,
  ScrollRestoration,
  useLoaderData,
  useNavigate,
} from "@remix-run/react";
import type { ReactNode } from "react";
import { RouterProvider, useLocale } from "react-aria-components";
import { AuthenticityTokenProvider } from "remix-utils/csrf/react";

import { useIsBot } from "#app/helpers/is-bot";
import { CSRF } from "#app/modules/csrf.server";
import { cn } from "#app/std/cn";
import "#app/tailwind.css";

export async function loader({ request, context }: LoaderFunctionArgs) {
  let url = new URL(request.url);

  let headers = new Headers();

  // Cross-Site Request Forgery
  let csrf = new CSRF(context);
  let [token, cookie] = await csrf.commitToken(request);
  if (cookie) headers.append("set-cookie", cookie);

  // No-JS
  let noJS = await context.noJS.validate(request);
  await context.noJS.save(noJS, headers);

  // Response
  return json({ token, noJS, locale: context.i18n.locale, url }, { headers });
}

export default withMetronome(function App() {
  let { token, noJS, url } = useLoaderData<typeof loader>();

  let noJSURL = new URL(url);
  noJSURL.searchParams.set("no-js", "true");

  return (
    <AuthenticityTokenProvider token={token}>
      {noJS ? null : (
        <noscript>
          <meta httpEquiv="refresh" content={\`0; url=\${noJSURL.toString()}\`} />
        </noscript>
      )}

      <Outlet />
    </AuthenticityTokenProvider>
  );
});

export function Layout({ children }: { children: ReactNode }) {
  let isBot = useIsBot();
  let navigate = useNavigate();

  let { locale, direction } = useLocale();

  return (
    <RouterProvider navigate={navigate}>
      <html
        lang={locale}
        dir={direction}
        className="bg-white text-neutral-950 dark:bg-black dark:text-neutral-50"
      >
        <head>
          <meta charSet="utf-8" />
          <meta name="viewport" content="width=device-width,initial-scale=1" />
          <Meta />
          <Links />
        </head>

        <body className={cn("min-h-screen w-screen")}>
          {children}
          {!isBot && (
            <>
              <ScrollRestoration />
              <Scripts />
            </>
          )}
        </body>
      </html>
    </RouterProvider>
  );
}
"
`;

exports[`transformRootTsx > should transform __templates__/complex-root.txt 2`] = `
SourceMap {
  "file": "complex-root.txt",
  "mappings": "AAAA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;",
  "names": [],
  "sources": [
    "complex-root.txt",
  ],
  "sourcesContent": [
    "import { withMetronome } from "metronome-sh/react";
import type { LoaderFunctionArgs } from "@remix-run/cloudflare";
import { json } from "@remix-run/cloudflare";
import {
  Links,
  Meta,
  Outlet,
  Scripts,
  ScrollRestoration,
  useLoaderData,
  useNavigate,
} from "@remix-run/react";
import type { ReactNode } from "react";
import { RouterProvider, useLocale } from "react-aria-components";
import { AuthenticityTokenProvider } from "remix-utils/csrf/react";

import { useIsBot } from "#app/helpers/is-bot";
import { CSRF } from "#app/modules/csrf.server";
import { cn } from "#app/std/cn";
import "#app/tailwind.css";

export async function loader({ request, context }: LoaderFunctionArgs) {
  let url = new URL(request.url);

  let headers = new Headers();

  // Cross-Site Request Forgery
  let csrf = new CSRF(context);
  let [token, cookie] = await csrf.commitToken(request);
  if (cookie) headers.append("set-cookie", cookie);

  // No-JS
  let noJS = await context.noJS.validate(request);
  await context.noJS.save(noJS, headers);

  // Response
  return json({ token, noJS, locale: context.i18n.locale, url }, { headers });
}

export default withMetronome(function App() {
  let { token, noJS, url } = useLoaderData<typeof loader>();

  let noJSURL = new URL(url);
  noJSURL.searchParams.set("no-js", "true");

  return (
    <AuthenticityTokenProvider token={token}>
      {noJS ? null : (
        <noscript>
          <meta httpEquiv="refresh" content={\`0; url=\${noJSURL.toString()}\`} />
        </noscript>
      )}

      <Outlet />
    </AuthenticityTokenProvider>
  );
});

export function Layout({ children }: { children: ReactNode }) {
  let isBot = useIsBot();
  let navigate = useNavigate();

  let { locale, direction } = useLocale();

  return (
    <RouterProvider navigate={navigate}>
      <html
        lang={locale}
        dir={direction}
        className="bg-white text-neutral-950 dark:bg-black dark:text-neutral-50"
      >
        <head>
          <meta charSet="utf-8" />
          <meta name="viewport" content="width=device-width,initial-scale=1" />
          <Meta />
          <Links />
        </head>

        <body className={cn("min-h-screen w-screen")}>
          {children}
          {!isBot && (
            <>
              <ScrollRestoration />
              <Scripts />
            </>
          )}
        </body>
      </html>
    </RouterProvider>
  );
}
",
  ],
  "version": 3,
}
`;

exports[`transformRootTsx > should transform __templates__/default-template-root.txt 1`] = `
"import { withMetronome } from "metronome-sh/react";
import { Links, Meta, Outlet, Scripts, ScrollRestoration } from "@remix-run/react";

export function Layout({ children }: { children: React.ReactNode }) {
  return (
    <html lang="en">
      <head>
        <meta charSet="utf-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1" />
        <Meta />
        <Links />
      </head>
      <body>
        {children}
        <ScrollRestoration />
        <Scripts />
      </body>
    </html>
  );
}

export default withMetronome(function App() {
  return <Outlet />;
});
"
`;

exports[`transformRootTsx > should transform __templates__/default-template-root.txt 2`] = `
SourceMap {
  "file": "default-template-root.txt",
  "mappings": "AAAA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;",
  "names": [],
  "sources": [
    "default-template-root.txt",
  ],
  "sourcesContent": [
    "import { withMetronome } from "metronome-sh/react";
import { Links, Meta, Outlet, Scripts, ScrollRestoration } from "@remix-run/react";

export function Layout({ children }: { children: React.ReactNode }) {
  return (
    <html lang="en">
      <head>
        <meta charSet="utf-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1" />
        <Meta />
        <Links />
      </head>
      <body>
        {children}
        <ScrollRestoration />
        <Scripts />
      </body>
    </html>
  );
}

export default withMetronome(function App() {
  return <Outlet />;
});
",
  ],
  "version": 3,
}
`;

exports[`transformRootTsx > should transform __templates__/simple-root.txt 1`] = `
"import { withMetronome } from "metronome-sh/react";
export default withMetronome(function App() {
  return <div>Hello, world!</div>;
});
"
`;

exports[`transformRootTsx > should transform __templates__/simple-root.txt 2`] = `
SourceMap {
  "file": "simple-root.txt",
  "mappings": "AAAA;AACA;AACA;AACA;",
  "names": [],
  "sources": [
    "simple-root.txt",
  ],
  "sourcesContent": [
    "import { withMetronome } from "metronome-sh/react";
export default withMetronome(function App() {
  return <div>Hello, world!</div>;
});
",
  ],
  "version": 3,
}
`;
